workspaces/cli-shared/src/captures/avro/file-system/ingest-s3-capture-saver.ts

```diff
@@ -0,0 +1,161 @@
+      Bucket: bucketName,
+      MaxKeys: 1000,
+      Prefix: prefix,
+      ContinuationToken: token
+    }).promise();
+    if (objects.Contents) {
+      let obj_keys = objects.Contents.map(o=>o.Key).filter((o): o is string => !!o);
+      if (objects.IsTruncated) {
+        let rest = await listAll(objects.NextContinuationToken);
+        return obj_keys.concat(rest);
+      } else {
+        return obj_keys;
+      }
+    } else {
+      console.log("No object contents?");
+      return [];
+    }
+  }
+
+  cli.action.start("Loading capture batch list");
+  let allKeys = await listAll();
+  cli.action.stop(`Found ${allKeys.length} capture batches to ingest`);
+  let objectsHandled = 0;
+  let passThrough = new stream.PassThrough();
+
+  const customBar = cli.progress({
+    format: 'Ingesting | {bar} | {value}/{total} batches',
+    barCompleteChar: '\u2588',
+    barIncompleteChar: '\u2591',
+  })
+  customBar.start(allKeys.length,0);
+
+  let fetchPump = async () => {
+    for (const key of allKeys) {
+      let objReadStream = s3.getObject({
+        Bucket: bucketName,
+        Key: key
+      }).createReadStream();
+      await pipeAsync(objReadStream, passThrough);
+      objectsHandled++;
+      customBar.update(objectsHandled);
+    }
+  }
+
+  let interactionCount = 0;
+
+  const oboePromose = new Promise<void>((resolve, reject) => {
+    oboe(passThrough)
+      .node('!', (row: IHttpInteraction) => {
+        if (!parsedRules.some(r => r.shouldIgnore(row.request.method, `${row.request.host}${row.request.path}?${row.request.query}`))) {
+          interactionCount++;
+          captureSaver.save(row);
+        }
+      })
+      .on('done', function () {
+        resolve();
+      })
+      .on('fail', function (e: any) {
+        console.error(e);
+        reject(e);
+      });
+  });
+
+  await Promise.all([fetchPump(), oboePromose]);
+
+  customBar.stop();
+
+  console.log(`Successfully loaded ${interactionCount} interaction events from your bucket!`);
+
+  const files = [
+    {
+      location: path.join(
+        capturesBaseDirectory,
+        captureId,
+        'optic-capture-state.json'
+      ),
+      contents: JSON.stringify({
+        captureId,
+        status: 'completed',
+        metadata: {
+          startedAt: new Date().toISOString(),
+          taskConfig: null,
+          lastInteraction: null,
+        },
+      }),
+    },
+  ];
+
+  await Promise.all(
+    files.map(async (x) => {
+      const { location, contents } = x;
+      await fs.ensureDir(path.dirname(location));
+      return fs.writeFile(location, contents);
+    })
+  );  
+
+  return interactionCount;
+}
+
+async function pipeAsync(tap: Readable, sink: Writable) {
+  return new Promise((resolve, reject) => {
+    // end: false is important -- it prevents the stream from ending when one source emits an `end` event
+    tap.pipe(sink, { end: false })
+    tap.on("end", resolve)
+    tap.on("error", reject)
+  })
+}
```

Comment:
This leaks memory the way the promise resolvement is implemented with event handlers here (no clean up of event listeners). You'll want to use `once` instead of `on` and also have a cleanup function removing the registered listeners upon 'end' or 'error', before resolving or rejecting. 

It also smells a bit, but possible through how it's named: connecting a pipe is a synchronous action, streams already an asynchronous construct. To "pipe" something asynchronously, when it's then actually done synchronously, seems non-sensical.

It seems what you're trying to do is create an 'StreamEnded' promise for the input stream, maybe something along those lines makes more sense.

Reply:
Considering this further, I think the smell goes beyond the name: having an existing stream and adding more and more sources to it is non-idiomatic. Generally, the approach with building stuff with streams is:

1. construct the pipeline synchronously
2. run until the stream at the tail end of the pipeline finishes, or any of them error out.

Reply:
Considering this further, I think the smell goes beyond the name: having an existing stream and adding more and more sources to it is non-idiomatic. Generally, the approach with building stuff with streams is:

1. construct the pipeline synchronously
2. run until the stream at the tail end of the pipeline finishes, or any of them error out.

\Considering this further, I think the smell goes beyond the name: having an existing stream and adding more and more sources to it is non-idiomatic. Generally, the approach with building stuff with streams is:

1. construct the pipeline synchronously
2. run until the stream at the tail end of the pipeline finishes, or any of them error out.

In this case, I would do this by:
1. create a stream of objects (from listing them)
2. map these objects to object read streams
3. merge these object read streams to a single read stream
4. Map the rows in the read streams to interactions that can be saved
5. map the interactions to the capture saver

Realising that Streams are Async Generators and Async Iterators now is the key to construct a pipeline like that in a reasonable way. Converting readable node streams to those early can make all that logic happen in a structured and typed way. See the `async-tools` and `streams` here for some ideas of how that's done: https://github.com/opticdev/optic/tree/develop/workspaces/diff-engine-wasm/src.

Also more than happy to pair with you on this, if it's worth addressing. The marriage between Node.js Streams and async / await / Promise has been a bit rough, until Async Generators and Async Iterators came to the scene, and it's not really widely talked about (I reckon since most Node engineers that would care are stuck in legacy stuff 😅).

Reply:
Refactored as per our call -- muuch nicer now